### PR TITLE
Feature/traceback serialization

### PR DESCRIPTION
**Pull request by ksamuel**
This is the first draft of what we talked about with the KISSest solution I can think of:

- store the serialized exception in the exception dict in `_result`
- add a `traceback` property to read it if needed

I had to make assumption about the `None` issue I mentioned in https://github.com/RealOrangeOne/django-tasks/pull/68#issuecomment-2225575756 since the branch is now rebased, force pushed and deleted. I resolved to return `None` from the traceback property in case of complex exceptions.

The PoC contains passing tests, including types and linting, but no documentation yet, awaiting your feedback to see if this is the spirit you are looking for.

I can squash if needed of course.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,82 @@
+{
```
Comment (RealOrangeOne) [inline .vscode/settings.json:1]: Issue: Please remove this file

File: tests/tests/test_utils.py [hunk 2]
```diff
@@ -96,16 +99,56 @@ def test_serialize_exceptions(self) -> None:
...
+                self.assertEqual(data["exc_traceback"], [dotted_path + msg])
+
+    def test_serialize_full_traceback(self) -> None:
+        try:
+            # Using optparse to generate an error because:
+            # - it's pure python
+            # - it's easy to trip down
+            # - it's unlikely to change ever
+            # - We can normalize the strack trace from it
+            optparse.OptionParser(option_list=[1])  # type: ignore
```
Comment (RealOrangeOne) [inline tests/tests/test_utils.py:128]: Suggestion: Why not just `1 / 0`?

File: django_tasks/task.py [hunk 3]
```diff
@@ -226,21 +227,34 @@ class TaskResult(Generic[T]):
...
-                exception_from_dict(cast(dict, self._result))
+                exception_from_dict(cast(SerializedExceptionDict, self._result))
                 if self._result is not None
                 else None
             )
 
         raise ValueError("Task has not finished yet")
 
+    @property
+    def traceback(self) -> Optional[Union[T, List[str]]]:
```
Comment (RealOrangeOne) [inline django_tasks/task.py:248]: I think this implementation can be simplified a little - If the task failed, then return it. Otherwise, be `None` - akin to `get_result` below

File: django_tasks/task.py [hunk 4]
```diff
@@ -226,21 +227,34 @@ class TaskResult(Generic[T]):
...
+        if self.status == ResultStatus.COMPLETE:
+            return None
+        if self.status == ResultStatus.FAILED:
+            if self._result is None:
+                return None
+            return cast(SerializedExceptionDict, self._result)["exc_traceback"]
```
Comment (RealOrangeOne) [inline django_tasks/task.py:254]: Question: Are there cases where there wouldn't be a traceback?

File: django_tasks/task.py (hunk 3 above)
Comment (RealOrangeOne) [inline django_tasks/task.py:248]: Suggestion: I think the traceback should be a single string, with the lines merged. Given it's mostly for debugging, that's going to be more useful for humans.

File: tests/tests/test_utils.py [hunk 5]
```diff
@@ -117,4 +160,4 @@ def test_cannot_deserialize_non_exception(self) -> None:
         ]:
             with self.subTest(data):
                 with self.assertRaises((TypeError, ImportError)):
-                    utils.exception_from_dict(data)
+                    utils.exception_from_dict(data)  # type: ignore
```
Comment (RealOrangeOne) [inline tests/tests/test_utils.py:163]: Suggestion: Remove the type ignore, and add an empty `exc_traceback` to the data fixtures.

File: tests/tests/test_utils.py [hunk 6]
```diff
@@ -96,16 +99,56 @@ def test_serialize_exceptions(self) -> None:
...
+            # Using optparse to generate an error because:
+            # - it's pure python
+            # - it's easy to trip down
+            # - it's unlikely to change ever
+            # - We can normalize the strack trace from it
+            optparse.OptionParser(option_list=[1])  # type: ignore
+        except Exception as e:
+            traceback = utils.exception_to_dict(e)["exc_traceback"]
+
+            expected_traceback_head = "Traceback (most recent call last):\n"
```
Comment (RealOrangeOne) [inline tests/tests/test_utils.py:132]: Nit: Defining these variables separately is unnecessary - just define them in the assertions.

Alternatively, why split them out at all? Just have 1 assertion.

**Review by RealOrangeOne — COMMENTED** [review]
Thanks so much for putting this together - it'll be a great boost for usability, without much extra maintenance burdern.

I'd also recommend rebasing your branch, now I've merged some more things in.

File: tests/tests/test_utils.py (hunk 2 above)
Comment (ksamuel) [reply tests/tests/test_utils.py:128]: It would tie the test to the code base, if the code base changes, the stack would change.

Since optparse is deprecated, but will likely stay in the stdlib for the next decade, it provides a stable stack trace to compare against for the test.

File: django_tasks/task.py (hunk 4 above)
Comment (ksamuel) [reply django_tasks/task.py:254]: If we didn't think of an edge case when producing the traceback, this means a bug on our side would not crash task retrieval, only provide no information.

I don't know if it's worth keeping. You choose.

File: tests/tests/test_utils.py (hunk 2 above)
Comment (RealOrangeOne) [reply tests/tests/test_utils.py:128]: I'd still expect the stacktrace to include the codebase, although I agree a deeper stack does help the test case a little more.

I think it'd be nice to have a flatter stack test too, perhaps also one which originates from C

File: tests/tests/test_utils.py (hunk 6 above)
Comment (ksamuel) [reply tests/tests/test_utils.py:132]: Because those parts of the traceback are constant, while there are a few lines that will depend on the setup. I'll add a comment about this.

File: django_tasks/task.py (hunk 4 above)
Comment (ksamuel) [reply django_tasks/task.py:254]: There is one thing to consider. In `test_complex_exception()`, the task is marked as failed yet `._result` is `None`. In that case, the task is marked as `FAILED`, but no dict is stored in `._result`:

```python
  def test_complex_exception(self) -> None:
        ....
        self.assertEqual(result.status, ResultStatus.FAILED)

        self.assertIsNone(result.result)
        self.assertIsNone(result.traceback) # this would fail if it assumes a dict is always there
        ...
  ```

So either I keep the check that the result is not `None`, or the behavior for complex exceptions must change.

File: tests/tests/test_utils.py (hunk 5 above)
Comment (ksamuel) [reply tests/tests/test_utils.py:163]: Not sure that will be sufficient for mypy, I might need to move the iterable into its own variable to declare the type explicitly.

**Comment by ksamuel** [comment]
Traceback are now strings, stylistic remarks are addressed, extranous file is removed. Typing is improved. Missing the C module stack trace for now.

File: django_tasks/task.py (hunk 4 above)
Comment (ksamuel) [reply django_tasks/task.py:256]: For now I kept the original behavior, let me know which option you prefer.

File: django_tasks/task.py (hunk 3 above)
Comment (ksamuel) [reply django_tasks/task.py]: This is dependent on the thread bellow.

File: tests/tests/test_utils.py [hunk 7]
```diff
@@ -101,20 +105,97 @@ def test_serialize_exceptions(self) -> None:
...
+            traceback = utils.exception_to_dict(e)["exc_traceback"]
+            self.assertTrue(traceback.startswith("Traceback (most recent call last):"))
+            self.assertTrue(
+                traceback.endswith(
+                    "TypeError: object supporting the buffer API required\n"
+                )
+            )
+            # Check that it's indeed a short traceback that sees mostly
+            # the error line
+            self.assertTrue("hashlib.md5(1)" in traceback)
```
Comment (RealOrangeOne) [inline tests/tests/test_utils.py:173]: Nit: 
```suggestion
            self.assertIn("hashlib.md5(1)", traceback)
```

File: django_tasks/task.py (hunk 4 above)
Comment (RealOrangeOne) [reply django_tasks/task.py:256]: Yeah if we can't serialize the exception, `None` is fine, which means the code here is fine. I think you're right there's not likely to be a case where there's an exception and no traceback.

File: django_tasks/task.py [hunk 8]
```diff
@@ -226,21 +226,37 @@ class TaskResult(Generic[T]):
...
 
         raise ValueError("Task has not finished yet")
 
+    @property
+    def traceback(self) -> Optional[str]:
+        """
+        Return the string representation of the traceback of the task if it failed
+        """
+
+        if self._result is None:
```
Comment (RealOrangeOne) [inline django_tasks/task.py:252]: Suggestion: Use a single case for the `None` return:

```python
if self._result is None or self.status != ResultStatus.FAILED:
```

**Review by RealOrangeOne — COMMENTED** [review]
Definitely the right direction :partying_face: Couple nits, but otherwise this is looking really nice!

File: django_tasks/task.py (hunk 8 above)
Comment (ksamuel) [reply django_tasks/task.py]: Done, but I inverted the logic, so that the specific working case return the value, otherwise, anything else returns None.

**Comment by ksamuel** [comment]
Nits addressed, linting passed. Once everything matches your expectations, I'll squash and force push.

**Comment by RealOrangeOne** [comment]
Looking at the current test failures, asserting the full depth of the stacktrace is likely to cause issues, and ties the tests to how the current version of Python implements it. Given `format_exception` is a built-in, I'm happy to trust that it works, and just throw some assertions in for the message itself and perhaps a file path or 2.

**Comment by ksamuel** [comment]
Indeed. Especially since error messages have been the focus of improvements by Pablo Galindo Salgado. I'll make the checks more fuzzy.

**Comment by ksamuel** [comment]
Better, now we test that the data looks like we put a traceback in there, but doesn't care much about the formatting of it. I also ran the sqlite test for good measure, although I don't expect the result store to be affected in any way.

**Comment by ksamuel** [comment]
Cheers, and thanks for all the work on a very useful addition to django.
